cpp/src/arrow/compute/kernel.h

```diff
@@ -55,68 +56,20 @@ class ARROW_EXPORT OpKernel {
   virtual ~OpKernel() = default;
 };
 
-/// \brief Placeholder for Scalar values until we implement these
-struct ARROW_EXPORT Scalar {
-  util::variant<bool, uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t,
-                int64_t, float, double>
-      value;
-
-  explicit Scalar(bool value) : value(value) {}
-  explicit Scalar(uint8_t value) : value(value) {}
-  explicit Scalar(int8_t value) : value(value) {}
-  explicit Scalar(uint16_t value) : value(value) {}
-  explicit Scalar(int16_t value) : value(value) {}
-  explicit Scalar(uint32_t value) : value(value) {}
-  explicit Scalar(int32_t value) : value(value) {}
-  explicit Scalar(uint64_t value) : value(value) {}
-  explicit Scalar(int64_t value) : value(value) {}
-  explicit Scalar(float value) : value(value) {}
-  explicit Scalar(double value) : value(value) {}
-
-  Type::type kind() const {
-    switch (this->value.which()) {
-      case 0:
-        return Type::BOOL;
-      case 1:
-        return Type::UINT8;
-      case 2:
-        return Type::INT8;
-      case 3:
-        return Type::UINT16;
-      case 4:
-        return Type::INT16;
-      case 5:
-        return Type::UINT32;
-      case 6:
-        return Type::INT32;
-      case 7:
-        return Type::UINT64;
-      case 8:
-        return Type::INT64;
-      case 9:
-        return Type::FLOAT;
-      case 10:
-        return Type::DOUBLE;
-      default:
-        return Type::NA;
-    }
-  }
-};
-
 /// \class Datum
 /// \brief Variant type for various Arrow C++ data structures
 struct ARROW_EXPORT Datum {
   enum type { NONE, SCALAR, ARRAY, CHUNKED_ARRAY, RECORD_BATCH, TABLE, COLLECTION };
 
-  util::variant<decltype(NULLPTR), Scalar, std::shared_ptr<ArrayData>,
+  util::variant<decltype(NULLPTR), std::shared_ptr<Scalar>, std::shared_ptr<ArrayData>,
```

Comment:
given scalars are supposed to be "small" it might be worth comment on why you use a shared_ptr type here (i.e. instead of the previous variant)

Reply:
* Scalars of nested types may be semantically small but physically occupy a fairly large amount of memory
* The object model uses subclasses. So you need to use a pointer type of some kind here